python/iceberg/parquet/arrow_table_utils.py

```diff
@@ -0,0 +1,203 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+
+from iceberg.api.expressions import Operation, Predicate
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+
+
+_logger = logging.getLogger(__name__)
+
+
+def apply_iceberg_filter(table, expected_to_file_map, expr):
+    """
+    Builds an indicator array to apply to a pandas dataframe and then returns the resulting
+    pyarrow.Table
+
+    Parameters
+    ----------
+    table: pyarrow.Table
+        The table to apply the filter on
+    expected_to_file_map: map
+        A mapping of the current schema field names to the file field names
+    expr: iceberg.api.expressions.Expression
+        A predicate expression to use as a filter
+
+    Returns
+    -------
+    pyarrow.Table
+        A pyarrow Table with the rows that don't match the predicate filtered out
+
+    """
+    if expr is None:
+        return table
+
+    indicator_array = get_indicator(table, expected_to_file_map, expr)
+    return pa.Table.from_pandas(table.to_pandas()[indicator_array], schema=table.schema)
```

Comment:
We actually have the C++ code for the filtering but not exposed it to Python, can you add a TODO reference to https://issues.apache.org/jira/browse/ARROW-6996 ?

Reply:
sure. I can add that.  Internally at Netflix we have a non-opensourced C library for doing row-level filtering.  I don't think we have any plans in the short-term for open-sourcing it, so it would be good to get anything available in arrow esp, if it's in C++ code.